src/analysis/skill_curve.rs

```diff
@@ -0,0 +1,223 @@
+
+            // Collect initial weighted segments
+            let mut current_weight = 1.0;
+            for &(id, time) in segment.segment_history().iter_actual_runs().rev() {
+                if let Some(time) = time[method] {
+                    // Skip all the combined segments
+                    let skip = catch! {
+                        segments[i.checked_sub(1)?].segment_history().get(id)?[method].is_none()
+                    }
+                    .unwrap_or(false);
+
+                    if !skip {
+                        weighted_segment_times.push((current_weight, time));
+                        current_weight *= WEIGHT;
+                    }
+                }
+            }
+
+            // End early if we don't have any segment times anymore
+            if weighted_segment_times.is_empty() {
+                len = i;
+                break;
+            }
+
+            // Sort everything by the times
+            weighted_segment_times
+                .sort_unstable_by_key(|&(_, time)| OrderedFloat(time.total_milliseconds()));
+
+            // Cumulative sum of the weights
+            let mut sum = 0.0;
+            for (weight, _) in weighted_segment_times.iter_mut() {
+                sum += *weight;
+                *weight = sum;
+            }
+
+            // Reweigh all of the weights to be in the range 0..1
+            let min = weighted_segment_times
+                .first()
+                .map(|&(w, _)| w)
+                .unwrap_or_default();
+
+            let max = weighted_segment_times
+                .last()
+                .map(|&(w, _)| w)
+                .unwrap_or_default();
+
+            let diff = max - min;
+
+            if diff != 0.0 {
+                for (weight, _) in weighted_segment_times.iter_mut() {
+                    *weight = (*weight - min) / diff;
+                }
+            }
+        }
+
+        // Limit the slice to only the segments that have segment times.
+        self.truncate(len);
+    }
+
+    /// This function returns an iterator that iterates over each segment and
+    /// yields their segment times for the percentile specified. A percentile of
+    /// 0 yields the fastest times, while a percentile of 1 yields the slowest
+    /// times.
+    pub fn iter_segment_times_at_percentile(
+        &self,
+        percentile: f64,
+    ) -> impl Iterator<Item = TimeSpan> + '_ {
+        self.all_weighted_segment_times
+            .iter()
+            .map(move |weighted_segment_times| {
+                let found_index = weighted_segment_times
+                    .binary_search_by(|&(w, _)| w.partial_cmp(&percentile).unwrap());
+
+                match found_index {
+                    // The percentile perfectly matched a segment time
+                    Ok(index) => weighted_segment_times[index].1,
+                    // The percentile didn't perfectly match, interpolate instead
+                    Err(right_index) => {
```

Comment:
Should we mention in the documentation for this function that you get an `Err` for any segment where you don't have an exact percentile match?

Reply:
You don't get an error. The curve is a linear interpolation of the individual segment times you got. Whether we need to interpolate or not is figured out by the binary search that is done. If it doesn't exactly find a match we have to interpolate. But that's almost always the case.

Reply:
Ok / Err is just the Rust standard library's way to signal the result of the binary search. https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search

Reply:
That makes sense, thanks for the clarification.